### DISCUSSION TITLE
Compile Query API

### DISCUSSION THREAD

**Discussion by nikitagashkov** [sqlalchemy/sqlalchemy#10997] [Ideas]
Hey!

First of all, thank you for the excellent library 🙂

I am currently exploring the available options to use SQLAlchemy as a pure query builder to directly feed the compiled result into the database driver.

It seems that the support for this could have been done rather simply if not for one thing — client-side defaults. Similar projects have implemented support for `default`/`onupdate`:

* asyncpgsa — https://github.com/CanopyTax/asyncpgsa/blob/0.25.0/asyncpgsa/connection.py#L27-L54
* aiopg — https://github.com/aio-libs/aiopg/blob/v1.4.0/aiopg/sa/engine.py#L19-L32
* asyncom — https://github.com/vinissimus/asyncom/blob/0.3.1/asyncom/om.py#L174-185
* databases — https://github.com/encode/databases/pull/206

Implementing such prefetching is rather tricky to get right due to the sheer variety of options available. None of the listed libraries support defaults to the full extent that SQLAlchemy itself does (e.g., no libraries support `insert_sentinel`s, only some support sequences).

So, the proposal is to either provide an API to execute defaults or to execute them during statement compilation. There are still some open questions:

1. What should be done with [context-sensitive default functions](https://docs.sqlalchemy.org/en/20/core/defaults.html#context-sensitive-default-functions)?
2. The current implementation of `_exec_default` depends on the connection and executes scalars prior to rendering the final prefetches. Is it possible to remove this dependency and fallback to pure compilation? For example, `default=Sequence("myseq")` can be compiled to `nextval('myseq')` without the need to execute it on the connection. Does this translate to all cases?..

Thanks!

**Comment by zzzeek** [comment]
> Implementing such prefetching is rather tricky to get right due to the sheer variety of options available. 

it's hard to do client side defaults, sure

> None of the listed libraries support defaults to the full extent that SQLAlchemy itself does 

at least one of those libraries is kind of a "uses sqlalchemy to make queries" kind of thing (asyncpgsa), that library is obsolete since we added async support

> (e.g., no libraries support insert_sentinels, only some support sequences).

an "insert sentinel" from our POV is a brand new thing we just added in the 2.0 series to help with a special use case of inserting lots of rows in bulk, using batched INSERT statements with many VALUES clauses, and the sentinel thing is to ensure we get RETURNING results that can be matched to the original VALUES clauses.  this is a totally new thing and SQLAlchemy had no such capabilitty from 2006-2023, so, not really a major thing if someone is writing a new library

> So, the proposal is to either provide an API to execute defaults or to execute them during statement compilation. There are still some open questions:

Huh ?   What are we talking about, SQLAlchemy compilation?  you can't "execute" SQL during compilation it just generates a string, and if a Python side default, you ceratinly dont make the value during string compilation.

if you want to execute a default, if it's in Python you just run the function, or if its a SQL expression just select() it from a connection.   If someone is writing a whole SQL compilation -> execute against the driver system, sure that is complicated, which is why....SQLAlchemy already does this.    


> The current implementation of _exec_default depends on the connection and executes scalars prior to rendering the final prefetches. Is it possible to remove this dependency and fallback to pure compilation? 

_exec_default is a private method.  if you like what it does and want it to do something different, copy it into your own library and make it do whatever you want (that's what [asyncpgsa is doing here](https://github.com/CanopyTax/asyncpgsa/blob/0.25.0/asyncpgsa/connection.py#L27-L54).

overall I am not entirely clicking with what you want to achieve.   the above mentioned async libraries are there either because SQLAlchemy did not used to support asyncio (and are hence unnecessary), or those libraries wanted to provide alternatives (which is fine, so they wrote the code to do what they needed).   If you want a system that uses SQLAlchemy to a. create SQL constructs b. compile them c. execute them against DBAPIs either sync or async, SQLAlchemy already does that, and if you want to see how involved that is, look at the source code!   if you want to write that yourself, then go ahead, but I dont understand the request "please crack open all kinds of SQLAlchemy internal functions so I can write my own library using 90% of SQLAlchemy's internals to do so" because what have you actually created?  taht's the question here
